File: 1448-count-good-nodes-in-binary-tree/1448-count-good-nodes-in-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        
        def dfs(root, M):
            if not root: return
            if root.val >= M:
                self.res += 1
            M = max(root.val, M)
            
            dfs(root.left, M)
            dfs(root.right, M)
        self.res = 0
        dfs(root, root.val)
        return self.res

File: 1448-count-good-nodes-in-binary-tree/1448-count-good-nodes-in-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int good(TreeNode* root, int parent) {
        if (not root)
            return 0;
        if (root->val >= parent)
            return 1 + good(root->left, root->val) + good(root->right, root->val);
        return good(root->left, parent) + good(root->right, parent);
    }
    int goodNodes(TreeNode* root) {
        return good(root, INT_MIN);
    }
};